### ISSUE TITLE
Using predict() with RcppArmadillo/RcppEigen when a factor level has only one level

### ISSUE THREAD

**Issue by oliviermeslin** [RcppCore/RcppArmadillo#303]
Note: I asked the same question on SO: [here](https://stackoverflow.com/questions/62282727/using-predict-with-rcpparmadillo-rcppeigen-when-a-factor-level-has-only-one-le).

I have a question regarding the use of the `predict()` function with the `RcppArmadillo` and `RcppEigen` packages, when a factor variable has only one level. I built a MWE below using the `iris` dataset.

I want to first estimate a linear regression model using `RcppArmadillo`, and then use it to predict values. The data I use for estimation contains factor variables (with more than one level and no `NA`). The prediction I want to make is slightly unusual in one respect: I want to predict values using the same factor level for all observations (this level being in levels used in the estimation). In the example below, it means that I want to predict `Sepal.Length` as if all observations were from the "versicolor" Species.

This works well when I estimate the model using the `lm()` function, but does not work with the `RcppArmadillo::fastLm()` or `RcppEigen::fastLm()` functions. I get the following error: ``Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : contrasts can be applied only to factors with 2 or more levels``. The same error happens again if one of the factor levels is missing.

 After some trial-and-error, I found this dirty work-around:

- I stack two versions of the data: the first one is the original data (with all factor levels), the second one is the modified data (with the same factor level for all observations).
- I predict values on this dataset; it works because all factor levels are present in the data;
- I keep only the modified subset of data.

Is this a bug, or am I using fastLm not properly?

```{r}
library(data.table)

# Loading iris data
iris <- as.data.table(iris)

# Estimating the model
model <-
  RcppArmadillo::fastLm(Sepal.Length ~ 
                 factor(Species)
               + Sepal.Width 
               + Petal.Length 
               + Petal.Width,
               data=iris)

summary(model)

#### 
#### Here is the error I don't understand
#### 

# This is the standard use of the predict function
iris2 <- copy(iris)
iris2[, predict := predict(model, iris2)]

# This is the way I want to use the predict function
# This does not work for some reason
iris2 <- copy(iris)
iris2[, Species := "versicolor"]
iris2[, predict2 := predict(model, iris2)]

#### 
#### This is a dirty work-around
#### 

# Creating a modified dataframe
iris3 <- copy(iris)
iris3[, `:=`(Species = "versicolor",
             data = "Modified data")]

# copying the original dataframe
iris4 <- copy(iris)
iris4[, data := "Original data"]

# Stacking the original data and the modified data
iris5 <- rbind(iris3, iris4)
iris5[, predict := predict(model, iris5)]

# Keeping only the modified data
iris_final <- iris5[data == "Modified data"]
```

**Comment by eddelbuettel** [comment]
What makes you think spamming us with crossposts is a good idea?

What makes you think I didn't already see your question on StackOverflow?

**Comment by eddelbuettel** [comment]
Aren't you simply up against this?  Quoting from `help(fastLm, package="RcppArmadillo")`:

> 
> Lastly, one must be be careful in timing comparisons of lm and friends versus this approach based on Armadillo. The reason that Armadillo can do something like lm.fit faster than the functions in the stats package is because Armadillo uses the Lapack version of the QR decomposition while the stats package uses a modified Linpack version. Hence Armadillo uses level-3 BLAS code whereas the stats package uses level-1 BLAS. However, Armadillo will either fail or, worse, produce completely incorrect answers on rank-deficient model matrices whereas the functions from the stats package will handle them properly due to the modified Linpack code.
> 
> An example of the type of situation requiring extra care in checking for rank deficiency is a two-way layout with missing cells (see the examples section). These cases require a special pivoting scheme of “pivot only on (apparent) rank deficiency” which is not part of conventional linear algebra software.

**Comment by oliviermeslin** [comment]
I'm really sorry if that wasn't the right way to proceed, I won't do it again.

I obviously read the help of the `fastLm()` function, but in my understanding the excerpt you mentioned is about the estimation process (a factor variable with only one level induces a rank deficient matrix), whereas my problem is about the `predict()` function. 

My question is then: Is there some matrix decomposition involved in the internals of the `predict()` method of `RcppArmadillo`, meaning you can't use it if some of the factor levels are not present in the dataset you want to produce predictions with?

Sorry again for bothering you.

**Comment by eddelbuettel** [comment]
It's open source and there for you to read.  If there is something that doesn't work in a corner case, you can a) report it (preferably only once), b) look into why it is that way, c) even try to correct it. 

I mostly just added formula notation to `fastLm()` as an exercise to see how the mechanics work. As I demonstrate in timings, _using a formula destroys all speed advantages from `fastLm`_ as so much time is spent parsing the formula and setting up the model matrix.  Hence, not sure how much I care about factor-level corner cases.   

Also, isn't a factor level of rank one the same as an additional intercept?  And I guess we all agree that you can't have two of those.  So something somewhere here isn;t quite right, and I haven't had time to think whether it is us, or your setup.  I somehow suspect the latter but I could be wrong.
